jina/types/document/multimodal.py

```diff
@@ -34,48 +31,51 @@ def __init__(self, document = None, chunks: Sequence[Document] = None, modality_
+        if chunks or modality_content_map:
             if chunks:
+                granularities = [chunk.granularity for chunk in chunks]
+                if len(set(granularities)) != 1:
+                    raise BadDocType('Each chunk should have the same granularity.')
                 self.chunks.extend(chunks)
-            if not chunks and modality_content_mapping:
-                self._add_chunks_from_modality_content_mapping(modality_content_mapping)
+            elif modality_content_map:
+                self.modality_content_map = modality_content_map
             self._handle_chunk_level_attributes()
-            self._validate()
-
-    def _build_modality_content_mapping(self) -> Dict:
-        for chunk in self.chunks:
-            modality = chunk.modality
-            self._modality_content_mapping[modality] = chunk.embedding \
-                if chunk.embedding is not None \
-                else chunk.content
-        self._validate()
 
-    def _add_chunks_from_modality_content_mapping(self, modality_content_mapping):
-        for modality, content in modality_content_mapping.items():
-            with Document() as chunk:
-                chunk.modality = modality
-                chunk.content = content
-                self.chunks.add(chunk)
+    @property
+    def is_valid(self) -> bool:
+        """A valid :class:`MultimodalDocument` should meet the following requirements:
 
-    def _validate(self):
+            - Document should consist at least 2 chunks.
+            - Length of modality is not identical to length of chunks.
+        """
         modalities = set([chunk.modality for chunk in self.chunks])
-        if len(self.chunks) < 2:
-            raise BadDocType('MultimodalDocument should consist at least 2 chunks.')
-        if len(modalities) != len(self.chunks):
-            raise LengthMismatchException(f'Length of modality is not identical to length of chunks.')
+        return 2 <= len(self.chunks) == len(modalities)
 
     def _handle_chunk_level_attributes(self):
         """Handle chunk attributes, such as :attr:`granularity` and :attr:`mime_type`.
 
         Chunk granularity should be greater than parent granularity level. Besides, if the chunk do not have
             a specified :attr:`mime_type`, it will be manually set to it's parent's :attr:`mime_type`.
         """
+
+        # Joan: https://github.com/jina-ai/jina/pull/1335#discussion_r533905780
+        # If chunk.granularity is 0. (This means a user without caring for granularity wants
+        #   to merge N documents into a multimodal document, therefore we do what
+        #   u have here of increasing their granularity inside this set) Well documented please
+        # If the chunk comes with granularity > 0, then it means that someone has cared to chunk already
+        #   the document or that we have some driver that generates muktimodal documents in the future.
+        #   Then, have document.granularity = chunk.granularity - 1.
```

Comment:
Maybe we need to clean this up

Reply:
I put some good thought on it and I truly believe it is the best trade off between coherency with granularity and ease of use from both Driver and Client side

Reply:
@bwanglzu did we put good unittest on this?

Reply:
@JoanFM yes we did